homeassistant/components/device_tracker/unifi.py

```diff
@@ -33,55 +40,87 @@
     vol.Required(CONF_USERNAME): cv.string,
     vol.Required(CONF_PORT, default=DEFAULT_PORT): cv.port,
     vol.Optional(CONF_VERIFY_SSL, default=DEFAULT_VERIFY_SSL): cv.boolean,
+    vol.Optional(CONF_DETECTION_TIME, default=DEFAULT_DETECTION_TIME): vol.All(
+        cv.time_period, cv.positive_timedelta),
+    vol.Optional(CONF_API_VERSION, default=DEFAULT_API_VERSION): cv.string
 })
 
 
-def get_scanner(hass, config):
+def get_scanner(hass, config, ectrl=None):
     """Set up the Unifi device_tracker."""
-    from pyunifi.controller import Controller, APIError
-
-    host = config[DOMAIN].get(CONF_HOST)
-    username = config[DOMAIN].get(CONF_USERNAME)
-    password = config[DOMAIN].get(CONF_PASSWORD)
-    site_id = config[DOMAIN].get(CONF_SITE_ID)
-    port = config[DOMAIN].get(CONF_PORT)
-    verify_ssl = config[DOMAIN].get(CONF_VERIFY_SSL)
-
+    from pyunifi.controller import APIError
+    config = config[DOMAIN]
+    result = None
+    if ectrl is not None:
+        _LOGGER.debug("ectrl found")
     try:
-        ctrl = Controller(host, username, password, port, version='v4',
```

Comment:
Why did you move this logic into the constructor ? It makes it more confusing what is going on .